keps/sig-storage/2485-read-write-once-pod-pv-access-mode/README.md

```diff
@@ -0,0 +1,972 @@
+lead to inefficient use of resources in their cluster. See [#30085] and [#26567]
+for issues related to this.
+
+[#30085]: https://github.com/kubernetes/kubernetes/issues/30085
+[#26567]: https://github.com/kubernetes/kubernetes/issues/26567
+
+### CSI Specification Changes
+
+In the CSI spec there are conflicting definitions of the [`SINGLE_NODE_WRITER`]
+access mode. By definition, `SINGLE_NODE_WRITER` means "Can only be published
+once as read/write on a single node, at any given time." The problem is how this
+access mode is used during `NodePublishVolume`, which is typically where volume
+mounting is performed.
+
+The CSI spec defines that when [`NodePublishVolume`] is called a second time for
+a volume with a non-`MULTI_NODE` access mode and with a different target path,
+the plugin should return `FAILED_PRECONDITION`. For CSI plugins that strictly
+adhere to the spec, this guarantees that a volume can only be mounted to a
+single target path, which means `SINGLE_NODE_WRITER` restricts access to a
+single pod on a single node. This behavior conflicts with the original
+definition. Due to this conflict, we do not have an access mode that represents
+multiple writers on the same node.
+
+[`SINGLE_NODE_WRITER`]: https://github.com/container-storage-interface/spec/blob/v1.4.0/csi.proto#L405-L407
+[`NodePublishVolume`]: https://github.com/container-storage-interface/spec/blob/v1.4.0/spec.md#nodepublishvolume
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Outline expected behavior of the ReadWriteOncePod access mode
+- Provide a high level design for ReadWriteOncePod access mode support
+- Define API changes needed to support this access mode
+- Outline changes needed in CSI spec and sidecars to support the
+  ReadWriteOncePod access mode
+- Outline changes needed in CSI spec and sidecars to continue supporting the
+  ReadWriteOnce access mode
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Kubernetes Changes, Access Mode
+
+In Kubernetes, we should add a new ReadWriteOncePod persistent volume access
+mode to PersistentVolumes and PersistentVolumeClaims. This change will require
+adding a feature gate and updating validation logic to accept this new access
+mode type if the feature gate is enabled.
+
+```golang
+       // can be mounted read/write mode to exactly 1 pod
+       ReadWriteOncePod PersistentVolumeAccessMode = "ReadWriteOncePod"
```

Comment:
Done, updated documentation on how this works. TL;DR: For kubelet and external-provisioner, it depends on the driver. For the attacher and resizer it will error.